add: effective degrees of freedom for penalized csplines

hlruh opened the pull request:
Adds option to compute the effective degrees of freedom. The effective degrees of freedom replace the classic degrees of freedom in the computation of the AIC and BIC information criteria for the pspline (Eilers & Marx, 1996). Computing the AIC or BIC allows to optimize the pspline penalty.

hlruh committed:
Update cspline.py

Adds option to compute the AIC and BIC information criteria using the effective degrees of freedom. Computing the AIC or BIC allows to optimize the pspline penalty. The effective degrees of freedom replace the classic degrees of freedom in the computation of the AIC and BIC for the pspline (Eilers & Marx, 1996).

hlruh committed:
Update cspline.py

Removed code unrelated to original commit.

hlruh committed:
Update cspline.py

Remove code unrelated to original commit.

hlruh committed:
Update cspline.py

Remove bug: naming of Nknot parameter.

mzechmeister commented on src/cspline.py:
```diff
@@ -409,8 +409,7 @@ def __call__(self, x=None):
 
-
-def ucbspl_fit(x, y=None, w=None, K=10, xmin=None, xmax=None, lam=0., pord=2, mu=None, e_mu=None, nat=True, retfit=False, var=False, e_yk=False, cov=False, plot=False, c=True):
+def ucbspl_fit(x, y=None, w=None, K=10, xmin=None, xmax=None, lam=0., pord=2, mu=None, e_mu=None, nat=True, retfit=False, var=False, e_yk=False, cov=False, ic= False, plot=False, c=True):
```
Remove space after `ic= `

hlruh replied:
replaced by edf

mzechmeister commented on src/cspline.py:
```diff
@@ -448,13 +447,15 @@ def ucbspl_fit(x, y=None, w=None, K=10, xmin=None, xmax=None, lam=0., pord=2, mu
    cov : boolean
       If false band matrices are used for efficient solution of equation system with band solver.
       If true covariances are estimated using matrix inversion.
+   ic : float, float, float
```
type is boolean

hlruh replied:
done

mzechmeister commented on src/cspline.py:
```diff
@@ -672,8 +677,115 @@ def ucbspl_fit(x, y=None, w=None, K=10, xmin=None, xmax=None, lam=0., pord=2, mu
+            lamDTD = lam*np.dot(D.T, D)
+                
+            # Compute the hat matrix for EDF
+            BT = B.T
```
Is a new variable `BT` needed?

hlruh replied:
not needed. removed.

mzechmeister commented on src/cspline.py:
```diff
@@ -672,8 +677,115 @@ def ucbspl_fit(x, y=None, w=None, K=10, xmin=None, xmax=None, lam=0., pord=2, mu
+                
+            # Compute the hat matrix for EDF
+            BT = B.T
+            BTWB = np.dot(np.dot(BT, W), B)
```
Swap lines and so BTW can be reused.

hlruh replied:
done.

mzechmeister commented on src/cspline.py:
```diff
@@ -672,8 +677,115 @@ def ucbspl_fit(x, y=None, w=None, K=10, xmin=None, xmax=None, lam=0., pord=2, mu
+            # effective degrees of freedom (edf)
+            edf = loop_sum_edf(BTWB_lamDTDinv, G, kk, w)
+            
+            # Compute BIC & AIC
```
move outside if-else so that the code block needs not to be duplicated

hlruh replied:
done.

mzechmeister commented on src/cspline.py:
```diff
@@ -672,8 +677,115 @@ def ucbspl_fit(x, y=None, w=None, K=10, xmin=None, xmax=None, lam=0., pord=2, mu
+    """
+       
+   # get matrix dimensions
+   Nknot, Ndat = BTWB_lamDTDinv.shape[0], G.shape[1]
```
Use two lines.

hlruh replied:
done.

mzechmeister commented on src/cspline.py:
```diff
@@ -672,8 +677,115 @@ def ucbspl_fit(x, y=None, w=None, K=10, xmin=None, xmax=None, lam=0., pord=2, mu
+   kn = kn.ravel()
+   ln = ln.ravel()
+
+   edf=0
```
space around `=`

hlruh replied:
done.

mzechmeister commented on src/cspline.py:
```diff
@@ -672,8 +677,115 @@ def ucbspl_fit(x, y=None, w=None, K=10, xmin=None, xmax=None, lam=0., pord=2, mu
+         # need to check if elements are valid for the last elements of GTGw
+         valid = np.flatnonzero((kki+kn < Nknot) & (kki+ln < Nknot))
+         if valid.size > 0:
+            # if elements are valid
```
Remove comment.

hlruh replied:
done.

hlruh committed:
Output only edf

Replaced ic parameter with edf. Return only edf, instead of information criteria aic, bic. Additional minor fixes.

mzechmeister committed:
Update cspline.py

fix: if statement.

hlruh committed:
Update cspline.py - fix: indention

fix: indention

mzechmeister committed:
style: Formatting.
